Template.yml errors result in stacktrace that is not readily apparent. Validation Enhancement?

Just a minor issue that can be ignored, but it would be nice to have a validation on the yml files and an error that is clearer. 

I had an error in my yml file where a space was missing. 
Exception did not so much warn that there was an erroneous config file but instead threw a jackson parsing exception. Additionally, this exception was due to a missing space after ":" which affects whether something is identified as a string or as a collection :grinning:  .

```
 2018-04-26 17:52:39.746 ERROR 50000 --- [eTaskExecutor-3] o.s.c.s.server.statemachine.ErrorAction  : Going through error state DefaultStateContext [stage=STATE_ENTRY, message=GenericMessage [payload=INSTALL, headers={INSTALL_REQUEST=org.springframework.cloud.skipper.domain.InstallRequest@780354c4, id=047727b4-b5c9-c614-756d-b80e0ab6448f, timestamp=1524757699262}], messageHeaders={INSTALL_REQUEST=org.springframework.cloud.skipper.domain.InstallRequest@780354c4, id=047727b4-b5c9-c614-756d-b80e0ab6448f, timestamp=1524757699262}, extendedState=DefaultExtendedState [variables={ERROR=com.fasterxml.jackson.databind.RuntimeJsonMappingException: Can not construct instance of java.util.LinkedHashMap: no String-argument constructor/factory method to deserialize from String value ('spring.cloud.deployer.memory:512m')
 at [Source: java.io.StringReader@3492392d; line: 15, column: 6] (through reference chain: org.springframework.cloud.skipper.domain.SpringCloudDeployerApplicationManifest["spec"]->org.springframework.cloud.skipper.domain.SpringCloudDeployerApplicationSpec["deploymentProperties"])}], transition=AbstractTransition [source=StateMachineState [getIds()=[INSTALL, INSTALL_INSTALL], toString()=AbstractState [id=INSTALL, pseudoState=null, deferred=[], entryActions=[], exitActions=[], stateActions=[], regions=[], submachine=INSTALL_INSTALL INSTALL_EXIT  /  / uuid=cfcb2543-c76b-41a7-ae8f-2d75131a801e / id=sg-decide-consultation-proxy-whatever3], getClass()=class org.springframework.statemachine.state.StateMachineState], target=ObjectState [getIds()=[ERROR_JUNCTION], getClass()=class org.springframework.statemachine.state.ObjectState, hashCode()=1990953100, toString()=AbstractState [id=ERROR_JUNCTION, pseudoState=org.springframework.statemachine.state.JunctionPseudoState@3373e5b2, deferred=[], entryActions=[], exitActions=[], stateActions=[], regions=[], submachine=null]], kind=EXTERNAL, guard=org.springframework.cloud.skipper.server.statemachine.StateMachineConfiguration$SkipperStateMachineFactoryConfig$$Lambda$13/4121176@5eb2972f], stateMachine=INSTALL INSTALL_INSTALL INSTALL_EXIT INITIAL ERROR ROLLBACK ROLLBACK_START ROLLBACK_EXIT_INSTALL ROLLBACK_EXIT ROLLBACK_EXIT_UPGRADE ROLLBACK_CHOICE UPGRADE UPGRADE_WAIT_TARGET_APPS UPGRADE_DEPLOY_TARGET_APPS_FAILED UPGRADE_DEPLOY_TARGET_APPS UPGRADE_CHECK_TARGET_APPS UPGRADE_START UPGRADE_CANCEL UPGRADE_DELETE_SOURCE_APPS UPGRADE_DEPLOY_TARGET_APPS_SUCCEED UPGRADE_EXIT UPGRADE_CHECK_CHOICE DELETE DELETE_DELETE DELETE_EXIT ERROR_JUNCTION  / ERROR / uuid=6a315d82-a8f4-42ac-9433-2fa5be971c76 / id=sg-decide-consultation-proxy-whatever3, source=null, target=null, sources=null, targets=null, exception=null] {}

com.fasterxml.jackson.databind.RuntimeJsonMappingException: Can not construct instance of java.util.LinkedHashMap: no String-argument constructor/factory method to deserialize from String value ('spring.cloud.deployer.memory:512m')
 at [Source: java.io.StringReader@3492392d; line: 15, column: 6] (through reference chain: org.springframework.cloud.skipper.domain.SpringCloudDeployerApplicationManifest["spec"]->org.springframework.cloud.skipper.domain.SpringCloudDeployerApplicationSpec["deploymentProperties"])
	at com.fasterxml.jackson.databind.MappingIterator.next(MappingIterator.java:194) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at org.springframework.cloud.skipper.domain.SpringCloudDeployerApplicationManifestReader.read(SpringCloudDeployerApplicationManifestReader.java:55) ~[spring-cloud-skipper-1.0.3.BUILD-SNAPSHOT.jar!/:1.0.3.BUILD-SNAPSHOT]
	at org.springframework.cloud.skipper.server.deployer.AppDeployerReleaseManager.install(AppDeployerReleaseManager.java:93) ~[spring-cloud-skipper-server-core-1.0.3.BUILD-SNAPSHOT.jar!/:1.0.3.BUILD-SNAPSHOT]
	at org.springframework.cloud.skipper.server.service.ReleaseService.install(ReleaseService.java:193) ~[spring-cloud-skipper-server-core-1.0.3.BUILD-SNAPSHOT.jar!/:1.0.3.BUILD-SNAPSHOT]
	at org.springframework.cloud.skipper.server.service.ReleaseService.install(ReleaseService.java:181) ~[spring-cloud-skipper-server-core-1.0.3.BUILD-SNAPSHOT.jar!/:1.0.3.BUILD-SNAPSHOT]
	at org.springframework.cloud.skipper.server.service.ReleaseService.install(ReleaseService.java:138) ~[spring-cloud-skipper-server-core-1.0.3.BUILD-SNAPSHOT.jar!/:1.0.3.BUILD-SNAPSHOT]
	at org.springframework.cloud.skipper.server.service.ReleaseService$$FastClassBySpringCGLIB$$f1c5f0a2.invoke(<generated>) ~[spring-cloud-skipper-server-core-1.0.3.BUILD-SNAPSHOT.jar!/:1.0.3.BUILD-SNAPSHOT]
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204) ~[spring-core-4.3.13.RELEASE.jar!/:4.3.13.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:738) ~[spring-aop-4.3.13.RELEASE.jar!/:4.3.13.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157) ~[spring-aop-4.3.13.RELEASE.jar!/:4.3.13.RELEASE]
	at org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99) ~[spring-tx-4.3.13.RELEASE.jar!/:4.3.13.RELEASE]
	at org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:282) ~[spring-tx-4.3.13.RELEASE.jar!/:4.3.13.RELEASE]
	at org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96) ~[spring-tx-4.3.13.RELEASE.jar!/:4.3.13.RELEASE]
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179) ~[spring-aop-4.3.13.RELEASE.jar!/:4.3.13.RELEASE]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:673) ~[spring-aop-4.3.13.RELEASE.jar!/:4.3.13.RELEASE]
	at org.springframework.cloud.skipper.server.service.ReleaseService$$EnhancerBySpringCGLIB$$8c3f8d4f.install(<generated>) ~[spring-cloud-skipper-server-core-1.0.3.BUILD-SNAPSHOT.jar!/:1.0.3.BUILD-SNAPSHOT]
	at org.springframework.cloud.skipper.server.statemachine.InstallInstallAction.executeInternal(InstallInstallAction.java:65) ~[spring-cloud-skipper-server-core-1.0.3.BUILD-SNAPSHOT.jar!/:1.0.3.BUILD-SNAPSHOT]
	at org.springframework.cloud.skipper.server.statemachine.AbstractAction.execute(AbstractAction.java:41) ~[spring-cloud-skipper-server-core-1.0.3.BUILD-SNAPSHOT.jar!/:1.0.3.BUILD-SNAPSHOT]
	at org.springframework.statemachine.state.AbstractState.executeAction(AbstractState.java:496) ~[spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.state.ObjectState.entry(ObjectState.java:158) ~[spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.AbstractStateMachine.entryToState(AbstractStateMachine.java:1288) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.AbstractStateMachine.entryToState(AbstractStateMachine.java:1232) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.AbstractStateMachine.setCurrentStateInternal(AbstractStateMachine.java:1038) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.AbstractStateMachine.setCurrentState(AbstractStateMachine.java:1012) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.AbstractStateMachine.switchToState(AbstractStateMachine.java:904) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.AbstractStateMachine.access$500(AbstractStateMachine.java:81) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.AbstractStateMachine$3.transit(AbstractStateMachine.java:332) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.DefaultStateMachineExecutor.handleInitialTrans(DefaultStateMachineExecutor.java:301) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.DefaultStateMachineExecutor.processTriggerQueueUntilConsumed(DefaultStateMachineExecutor.java:405) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.DefaultStateMachineExecutor.processTriggerQueue(DefaultStateMachineExecutor.java:387) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.DefaultStateMachineExecutor.access$200(DefaultStateMachineExecutor.java:65) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.DefaultStateMachineExecutor$1.run(DefaultStateMachineExecutor.java:329) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.core.task.SyncTaskExecutor.execute(SyncTaskExecutor.java:50) [spring-core-4.3.13.RELEASE.jar!/:4.3.13.RELEASE]
	at org.springframework.statemachine.support.DefaultStateMachineExecutor.scheduleEventQueueProcessing(DefaultStateMachineExecutor.java:352) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.DefaultStateMachineExecutor.execute(DefaultStateMachineExecutor.java:163) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.DefaultStateMachineExecutor.doStart(DefaultStateMachineExecutor.java:175) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.LifecycleObjectSupport.start(LifecycleObjectSupport.java:120) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.AbstractStateMachine.doStart(AbstractStateMachine.java:416) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.LifecycleObjectSupport.start(LifecycleObjectSupport.java:120) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.state.StateMachineState.entry(StateMachineState.java:226) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.AbstractStateMachine.entryToState(AbstractStateMachine.java:1288) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.AbstractStateMachine.entryToState(AbstractStateMachine.java:1232) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.AbstractStateMachine.setCurrentStateInternal(AbstractStateMachine.java:1038) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.AbstractStateMachine.setCurrentState(AbstractStateMachine.java:1012) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.AbstractStateMachine.switchToState(AbstractStateMachine.java:904) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.AbstractStateMachine.access$500(AbstractStateMachine.java:81) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.AbstractStateMachine$3.transit(AbstractStateMachine.java:335) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.DefaultStateMachineExecutor.handleTriggerTrans(DefaultStateMachineExecutor.java:286) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.DefaultStateMachineExecutor.handleTriggerTrans(DefaultStateMachineExecutor.java:211) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.DefaultStateMachineExecutor.processTriggerQueueUntilConsumed(DefaultStateMachineExecutor.java:457) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.DefaultStateMachineExecutor.processTriggerQueue(DefaultStateMachineExecutor.java:387) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.DefaultStateMachineExecutor.access$200(DefaultStateMachineExecutor.java:65) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at org.springframework.statemachine.support.DefaultStateMachineExecutor$1.run(DefaultStateMachineExecutor.java:323) [spring-statemachine-core-1.2.10.RELEASE.jar!/:1.2.10.RELEASE]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [na:1.8.0_161]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [na:1.8.0_161]
	at java.lang.Thread.run(Unknown Source) [na:1.8.0_161]
Caused by: com.fasterxml.jackson.databind.JsonMappingException: Can not construct instance of java.util.LinkedHashMap: no String-argument constructor/factory method to deserialize from String value ('spring.cloud.deployer.memory:512m')
 at [Source: java.io.StringReader@3492392d; line: 15, column: 6] (through reference chain: org.springframework.cloud.skipper.domain.SpringCloudDeployerApplicationManifest["spec"]->org.springframework.cloud.skipper.domain.SpringCloudDeployerApplicationSpec["deploymentProperties"])
	at com.fasterxml.jackson.databind.JsonMappingException.from(JsonMappingException.java:270) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.DeserializationContext.instantiationException(DeserializationContext.java:1456) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.DeserializationContext.handleMissingInstantiator(DeserializationContext.java:1012) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.deser.ValueInstantiator._createFromStringFallbacks(ValueInstantiator.java:370) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator.createFromString(StdValueInstantiator.java:315) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:355) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.deser.std.MapDeserializer.deserialize(MapDeserializer.java:27) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:504) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:104) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:276) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:140) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.deser.SettableBeanProperty.deserialize(SettableBeanProperty.java:504) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:104) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:276) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:140) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.MappingIterator.nextValue(MappingIterator.java:277) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	at com.fasterxml.jackson.databind.MappingIterator.next(MappingIterator.java:192) ~[jackson-databind-2.8.10.jar!/:2.8.10]
	... 55 common frames omitted
```

And the difference between a correct and incorrect file was:

```
  deploymentProperties:
     spring.cloud.deployer.memory: {{sgDeployerMemory}}
```
**Note: Correct version has a space after :**

vs

```
  deploymentProperties:
     spring.cloud.deployer.memory:{{sgDeployerMemory}}
```

--- a/spring-cloud-skipper/src/main/java/org/springframework/cloud/skipper/domain/SpringCloudDeployerApplicationManifestReader.java
+++ b/spring-cloud-skipper/src/main/java/org/springframework/cloud/skipper/domain/SpringCloudDeployerApplicationManifestReader.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2017 the original author or authors.
+ * Copyright 2017-2018 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -25,6 +25,7 @@ import java.util.Map;
 import com.fasterxml.jackson.databind.DeserializationFeature;
 import com.fasterxml.jackson.databind.JsonMappingException;
 import com.fasterxml.jackson.databind.MappingIterator;
+import com.fasterxml.jackson.databind.RuntimeJsonMappingException;
 import com.fasterxml.jackson.dataformat.yaml.YAMLMapper;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -59,6 +60,11 @@ public class SpringCloudDeployerApplicationManifestReader implements SkipperMani
 				}
 			}
 			catch (JsonMappingException e) {
+				logger.error("Can't parse Package's manifest YAML = " + manifest);
+				throw new SkipperException("JsonMappingException - Can't parse Package's manifest YAML = " + manifest,
+						e);
+			}
+			catch (RuntimeJsonMappingException e) {
 				logger.error("Can't parse Package's manifest YAML = " + manifest);
 				throw new SkipperException("JsonMappingException - Can't parse Package's manifest YAML = " + manifest,
 						e);

